### ISSUE TITLE
Circular Dependency Issues. Is this a compatibility issue?

### ISSUE THREAD

**Issue by applecrusher** [gottfrois/dashing-rails#64]
I can't seem to get a complete page loaded. Would anyone know what could be causing it to crash or what dependencies may be causing it to crash?

RuntimeError (Circular dependency detected while autoloading constant Dashing::WidgetsController):
  activesupport (4.1.8) lib/active_support/dependencies.rb:478:in `load_missing_constant'
  activesupport (4.1.8) lib/active_support/dependencies.rb:180:in`const_missing'
  activesupport (4.1.8) lib/active_support/inflector/methods.rb:240:in `const_get'
  activesupport (4.1.8) lib/active_support/inflector/methods.rb:240:in`block in constantize'
  activesupport (4.1.8) lib/active_support/inflector/methods.rb:236:in `each'
  activesupport (4.1.8) lib/active_support/inflector/methods.rb:236:in`inject'
  activesupport (4.1.8) lib/active_support/inflector/methods.rb:236:in `constantize'
  activesupport (4.1.8) lib/active_support/dependencies.rb:552:in`get'
  activesupport (4.1.8) lib/active_support/dependencies.rb:583:in `constantize'
  actionpack (4.1.8) lib/action_dispatch/routing/route_set.rb:78:in`controller_reference'
  actionpack (4.1.8) lib/action_dispatch/routing/route_set.rb:68:in `controller'
  actionpack (4.1.8) lib/action_dispatch/routing/route_set.rb:46:in`call'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:73:in `block in call'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:59:in`each'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:59:in `call'
  actionpack (4.1.8) lib/action_dispatch/routing/route_set.rb:678:in`call'
  railties (4.1.8) lib/rails/engine.rb:514:in `call'
  railties (4.1.8) lib/rails/railtie.rb:194:in`public_send'
  railties (4.1.8) lib/rails/railtie.rb:194:in `method_missing'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:73:in`block in call'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:59:in `each'
  actionpack (4.1.8) lib/action_dispatch/journey/router.rb:59:in`call'
  actionpack (4.1.8) lib/action_dispatch/routing/route_set.rb:678:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  rack-pjax (0.8.0) lib/rack/pjax.rb:12:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  omniauth (1.2.2) lib/omniauth/strategy.rb:186:in `call!'
  omniauth (1.2.2) lib/omniauth/strategy.rb:164:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  omniauth (1.2.2) lib/omniauth/strategy.rb:186:in`call!'
  omniauth (1.2.2) lib/omniauth/strategy.rb:164:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  omniauth (1.2.2) lib/omniauth/builder.rb:59:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/rack/agent_hooks.rb:30:in `traced_call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/rack/browser_monitoring.rb:23:in `traced_call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:55:in`call'
  warden (1.2.3) lib/warden/manager.rb:35:in `block in call'
  warden (1.2.3) lib/warden/manager.rb:34:in`catch'
  warden (1.2.3) lib/warden/manager.rb:34:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  rack (1.5.2) lib/rack/etag.rb:23:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  rack (1.5.2) lib/rack/conditionalget.rb:25:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  rack (1.5.2) lib/rack/head.rb:11:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  remotipart (1.2.1) lib/remotipart/middleware.rb:27:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  actionpack (4.1.8) lib/action_dispatch/middleware/params_parser.rb:27:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  actionpack (4.1.8) lib/action_dispatch/middleware/flash.rb:254:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  rack (1.5.2) lib/rack/session/abstract/id.rb:225:in `context'
  rack (1.5.2) lib/rack/session/abstract/id.rb:220:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/cookies.rb:560:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  activerecord (4.1.8) lib/active_record/query_cache.rb:36:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  activerecord (4.1.8) lib/active_record/connection_adapters/abstract/connection_pool.rb:621:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  activerecord (4.1.8) lib/active_record/migration.rb:380:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/callbacks.rb:29:in`block in call'
  activesupport (4.1.8) lib/active_support/callbacks.rb:82:in `run_callbacks'
  actionpack (4.1.8) lib/action_dispatch/middleware/callbacks.rb:27:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/reloader.rb:73:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/remote_ip.rb:76:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  airbrake (3.2.1) lib/airbrake/rails/middleware.rb:13:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/debug_exceptions.rb:17:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/show_exceptions.rb:30:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  railties (4.1.8) lib/rails/rack/logger.rb:38:in`call_app'
  railties (4.1.8) lib/rails/rack/logger.rb:22:in `call'
  quiet_assets (1.1.0) lib/quiet_assets.rb:27:in`call_with_quiet_assets'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/request_id.rb:21:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  rack (1.5.2) lib/rack/methodoverride.rb:21:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  rack (1.5.2) lib/rack/runtime.rb:17:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  activesupport (4.1.8) lib/active_support/cache/strategy/local_cache_middleware.rb:26:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  actionpack (4.1.8) lib/action_dispatch/middleware/static.rb:84:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  rack (1.5.2) lib/rack/sendfile.rb:112:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  airbrake (3.2.1) lib/airbrake/user_informer.rb:16:in`_call'
  airbrake (3.2.1) lib/airbrake/user_informer.rb:12:in `call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in`call'
  railties (4.1.8) lib/rails/engine.rb:514:in `call'
  railties (4.1.8) lib/rails/application.rb:144:in`call'
  newrelic_rpm (3.9.8.273) lib/new_relic/agent/instrumentation/middleware_tracing.rb:57:in `call'
  rack (1.5.2) lib/rack/content_length.rb:14:in`call'
  puma (2.10.2) lib/puma/server.rb:492:in `handle_request'
  puma (2.10.2) lib/puma/server.rb:363:in`process_client'
  puma (2.10.2) lib/puma/server.rb:254:in `block in run'
  puma (2.10.2) lib/puma/thread_pool.rb:101:in`call'
  puma (2.10.2) lib/puma/thread_pool.rb:101:in `block in spawn_thread'

**Comment by joseschmidt** [comment]
I think i'm getting the same problem, just did a bundle update and getting this when killing the application after no response:

    Puma caught this error: Attempt to unlock a mutex which is not locked (ThreadError)
    /home/warch/.rvm/gems/ruby-2.1.0/gems/rack-1.5.2/lib/rack/lock.rb:22:in `unlock'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/rack-1.5.2/lib/rack/lock.rb:22:in `ensure in call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/rack-1.5.2/lib/rack/lock.rb:23:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/actionpack-4.1.4/lib/action_dispatch/middleware/static.rb:64:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/rack-1.5.2/lib/rack/sendfile.rb:112:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/railties-4.1.4/lib/rails/engine.rb:514:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/railties-4.1.4/lib/rails/application.rb:144:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/rack-1.5.2/lib/rack/content_length.rb:14:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/puma-2.11.1/lib/puma/server.rb:507:in `handle_request'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/puma-2.11.1/lib/puma/server.rb:375:in `process_client'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/puma-2.11.1/lib/puma/server.rb:262:in `block in run'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `call'
    /home/warch/.rvm/gems/ruby-2.1.0/gems/puma-2.11.1/lib/puma/thread_pool.rb:104:in `block in spawn_thread'

**Comment by gottfrois** [comment]
Hum, i'll have a look, building a new app from scratch. What ruby and rails version are you using ?

**Comment by joseschmidt** [comment]
rails (4.1.4)


    $ bundle list
    Gems included by the bundle:
      * actionmailer (4.1.4)
      * actionpack (4.1.4)
      * actionview (4.1.4)
      * activemodel (4.1.4)
      * activerecord (4.1.4)
      * activesupport (4.1.4)
      * arel (5.0.1.20140414130214)
      * builder (3.2.2)
      * bundler (1.7.6)
      * coffee-rails (4.0.1)
      * coffee-script (2.3.0)
      * coffee-script-source (1.9.1)
      * connection_pool (2.1.2)
      * dashing-rails (2.4.3)
      * erubis (2.7.0)
      * execjs (2.4.0)
      * hike (1.2.3)
      * i18n (0.7.0)
      * jbuilder (2.2.11)
      * jquery-rails (3.1.2)
      * json (1.8.2)
      * mail (2.5.4)
      * mime-types (1.25.1)
      * minitest (5.5.1)
      * multi_json (1.11.0)
      * polyglot (0.3.5)
      * puma (2.11.1)
      * rack (1.5.2)
      * rack-test (0.6.3)
      * rails (4.1.4)
      * railties (4.1.4)
      * rake (10.4.2)
      * rdoc (4.2.0)
      * redis (3.2.1)
      * rufus-scheduler (3.0.9)
      * sass (3.2.19)
      * sass-rails (4.0.5)
      * sdoc (0.4.1)
      * spring (1.3.3)
      * sprockets (2.12.3)
      * sprockets-rails (2.2.4)
      * sqlite3 (1.3.10)
      * thor (0.19.1)
      * thread_safe (0.3.5)
      * tilt (1.4.1)
      * treetop (1.4.15)
      * turbolinks (2.5.3)
      * tzinfo (1.2.2)
      * uglifier (2.7.1)

**Comment by joseschmidt** [comment]
So, i don't understand this application very well. I'm getting no errors now and everything is running

**Comment by gottfrois** [comment]
multi threaded servers like Puma are quite new and there are still a lot of mystery around them unfortunately
